src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,551 @@
+            // Move scalar from base field into scalar field (which always fits
+            // for Pallas).
+            let scalar = pallas::Scalar::from_bytes(&scalar_val.to_bytes()).unwrap();
+            let expected = Point::new(
+                chip,
+                layouter.namespace(|| "expected point"),
+                Some((base_val * scalar).to_affine()),
+            )?;
+            result.constrain_equal(layouter.namespace(|| "constrain result"), &expected)
+        }
+
+        // [a]B
+        {
+            let scalar_val = pallas::Base::rand();
+            let result = {
+                let scalar = ScalarVar::new(
+                    chip.clone(),
+                    layouter.namespace(|| "random scalar"),
+                    Some(scalar_val),
+                )?;
+                p.mul(layouter.namespace(|| "random [a]B"), &scalar)?
+            };
+            constrain_equal(
+                chip.clone(),
+                layouter.namespace(|| "random [a]B"),
+                p_val,
+                scalar_val,
+                result,
+            )?;
+        }
+
+        // [a]𝒪 should return an error since variable-base scalar multiplication
+        // uses incomplete addition at the beginning of its double-and-add.
+        {
+            let scalar_val = pallas::Base::rand();
+            let scalar = ScalarVar::new(
+                chip.clone(),
+                layouter.namespace(|| "random scalar"),
+                Some(scalar_val),
+            )?;
+            zero.mul(layouter.namespace(|| "[a]𝒪"), &scalar)
+                .expect_err("[a]𝒪 should return an error");
+        }
+
+        // [0]B should return (0,0) since variable-base scalar multiplication
+        // uses complete addition for the final bits of the scalar.
+        {
+            let scalar_val = pallas::Base::zero();
+            let result = {
+                let scalar = ScalarVar::new(
+                    chip.clone(),
+                    layouter.namespace(|| "zero"),
+                    Some(scalar_val),
+                )?;
+                p.mul(layouter.namespace(|| "[0]B"), &scalar)?
+            };
+            constrain_equal(
+                chip.clone(),
+                layouter.namespace(|| "[0]B"),
+                p_val,
+                scalar_val,
+                result,
+            )?;
+        }
+
+        // [-1]B (the largest possible base field element)
+        {
+            let scalar_val = -pallas::Base::one();
+            let result = {
+                let scalar =
+                    ScalarVar::new(chip.clone(), layouter.namespace(|| "-1"), Some(scalar_val))?;
+                p.mul(layouter.namespace(|| "[-1]B"), &scalar)?
+            };
+            constrain_equal(
+                chip,
+                layouter.namespace(|| "[-1]B"),
+                p_val,
+                scalar_val,
+                result,
+            )?;
+        }
+
+        Ok(())
+    }
```

Comment:
Test the case where the last step (the complete addition) is a doubling. I think this occurs for the scalar 0x1b6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6e; please check this.

Reply:
I have checked that the last step is a doubling for 0x16db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6db6dc.